File: python/secret-handshake/secret_handshake.py


CODE = [
    'wink',
    'double blink',
    'close your eyes',
    'jump',
]
REVERSE = len(CODE)


def handshake(code):
    try:
        code = _as_int_code(code)
    except ValueError:
        return []

    codes = [i for i in range(len(CODE)) if (1 << i) & code]
    if (1 << REVERSE) & code:
        codes.reverse()

    return [CODE[c] for c in codes]


def code(words):
    try:
        codes = [CODE.index(w) for w in words]
    except ValueError:
        return '0'

    if codes != sorted(codes):
        codes.append(len(CODE))

    return '{:b}'.format(sum(1 << c for c in codes))


def _as_int_code(code):
    if isinstance(code, str):
        return int(code, 2)
    elif code < 0:
        raise ValueError('Code is negative')
    else:
        return code

File: go/secret-handshake/secret_handshake.go
package secret

const testVersion = 1

var handshakes = [...]string{
	"wink",
	"double blink",
	"close your eyes",
	"jump",
}

func Handshake(code uint) []string {
	var result = make([]string, 0)
	for i := 0; i < len(handshakes); i++ {
		if 1<<uint(i)&code != 0 {
			result = append(result, handshakes[i])
		}
	}
	if 1 << uint(len(handshakes)) & code != 0 {
		reverse(result)
	}
	return result
}

func reverse(slice []string) () {
	for i := 0; i < len(slice)/2; i++ {
		j := len(slice) - i - 1
		slice[i], slice[j] = slice[j], slice[i]
	}
}
